apps/server/src/modules/files-storage/service/files-storage.service.ts

```diff
@@ -123,9 +213,84 @@ export class FilesStorageService {
 	public async restore(fileRecords: FileRecord[]) {
 		this.logger.debug({ action: 'restore', fileRecords });
 
-		const unmarkFileRecords = this.filesStorageHelper.unmarkForDelete(fileRecords);
+		const unmarkFileRecords = unmarkForDelete(fileRecords);
 		await this.fileRecordRepo.save(unmarkFileRecords);
 
 		await this.restoreWithRollbackByError(fileRecords);
 	}
+
+	// copy
+	public async copyFilesOfParent(
+		userId: string,
+		params: FileRecordParams,
+		copyFilesParams: CopyFilesOfParentParams
+	): Promise<Counted<CopyFileResponse[]>> {
+		const [fileRecords, count] = await this.fileRecordRepo.findBySchoolIdAndParentId(params.schoolId, params.parentId);
+
+		if (count === 0) {
+			return [[], 0];
+		}
+
+		const response = await this.copy(userId, fileRecords, copyFilesParams.target);
+
+		return [response, count];
+	}
+
+	public async copyFileRecord(
+		sourceFile: FileRecord,
+		targetParams: FileRecordParams,
+		userId: EntityId
+	): Promise<FileRecord> {
+		const entity = createFileRecord(sourceFile.name, sourceFile.size, sourceFile.mimeType, targetParams, userId);
+
+		entity.securityCheck = deriveStatusFromSource(sourceFile, entity);
+
+		await this.fileRecordRepo.save(entity);
+
+		return entity;
+	}
+
+	public sendToAntiVirusService(sourceFile: FileRecord) {
+		if (sourceFile.securityCheck.status === ScanStatus.PENDING) {
+			this.antivirusService.send(sourceFile);
+		}
+	}
+
+	public async tryCopyFiles(sourceFile: FileRecord, targetFile: FileRecord): Promise<CopyFileResponse> {
+		try {
+			const paths = createICopyFiles(sourceFile, targetFile);
+
+			await this.storageClient.copy([paths]);
+
+			this.sendToAntiVirusService(sourceFile);
+
+			return new CopyFileResponse({ id: targetFile.id, sourceId: sourceFile.id, name: targetFile.name });
+		} catch (error) {
+			await this.fileRecordRepo.delete([targetFile]);
+			throw error;
+		}
+	}
+
+	public async copy(
+		userId: EntityId,
+		sourceFileRecords: FileRecord[],
+		targetParams: FileRecordParams
+	): Promise<CopyFileResponse[]> {
+		this.logger.debug({ action: 'copy', sourceFileRecords, targetParams });
+		const responseEntities: CopyFileResponse[] = [];
+
+		const promises = sourceFileRecords.map(async (sourceFile) => {
+			if (isStatusBlocked(sourceFile) || sourceFile.deletedSince) return;
+
+			const targetFile = await this.copyFileRecord(sourceFile, targetParams, userId);
+
+			const fileResponse = await this.tryCopyFiles(sourceFile, targetFile);
+
+			responseEntities.push(fileResponse);
+		});
+
+		await Promise.allSettled(promises);
```

Comment:
We should pass the information if some files are not copied. This is a important information for the share and copy service. But we can do it in next Ticket.

Reply:
Has seperate ticket https://ticketsystem.dbildungscloud.de/browse/BC-2459